### PR TITLE
Adding SkyPilot example for FlexGen

### PR DESCRIPTION
**Pull request by Michaelvll**
This PR is to add the SkyPilot example for the FlexGen benchmark. It will make the benchmark more reproducible and convenient to manage.

Several future TODOs for SkyPilot:
1. Use the `memory` filtering in the `resources` section (https://github.com/skypilot-org/skypilot/pull/1746) to make the example easier to run on different clouds, i.e.
```
resources:
  cpus: 32+
  memory_gb: 200+
  accelerators: T4
```
2. Add the support in changing the disk type to be used for the instance, so that we can run the commands that requires high performance SSD disks.

Tested:
- [x] `sky launch -c flexgen --use-spot --detach-setup ./flexgen/apps/task.yaml`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -66,9 +66,19 @@ python3 -m flexgen.apps.helm_run --description mmlu:model=text,subject=abstract_
 ```
 Note that only a subset of HELM scenarios is tested. See more tested scenarios [here](flexgen/apps/helm_passed_30b.sh).
 
+### Run FlexGen on any cloud with SkyPilot
```
Comment (concretevitamin) [inline README.md]: nit: Capital Case to be consistent with this doc?

File: README.md [hunk 2]
```diff
@@ -66,9 +66,19 @@ python3 -m flexgen.apps.helm_run --description mmlu:model=text,subject=abstract_
...
+FlexGen benchmark can be launched with [SkyPilot](http://skypilot.co), a tool for launching ML jobs on any cloud.
```
Comment (concretevitamin) [inline README.md]:
```suggestion
FlexGen benchmark can be launched with [SkyPilot](https://github.com/skypilot-org/skypilot), a tool for launching ML jobs on any cloud.
```

File: README.md [hunk 3]
```diff
@@ -66,9 +66,19 @@ python3 -m flexgen.apps.helm_run --description mmlu:model=text,subject=abstract_
...
+You can use a single command below to automatically launch the benchmark on any cloud with SkyPilot, after you setup your cloud account locally (check how to setup SkyPilot [here](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html)).
+```
+sky launch -c flexgen --detach-setup flexgen/apps/task.yaml
+```
+Note that you can replace the run section with any FlexGen command. You can log into the cluster running the job with `ssh flexgen` and terminate the cluster with `sky down flexgen`.
```
Comment (concretevitamin) [inline README.md]:
```suggestion
You can then log into the cluster running the job with `ssh flexgen` for monitoring. Once the job has finished, the cluster will be automatically terminated due to the `--down` flag.

To run any other FlexGen command, you can edit [`flexgen/apps/task.yaml`](./flexgen/apps/task.yaml) and replace the `run` section.
```

File: flexgen/apps/task.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
+# benchmark.yaml
```
Comment (concretevitamin) [inline flexgen/apps/task.yaml]: L1 can be removed

File: flexgen/apps/task.yaml [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+# A SkyPilot job definition for benchmarking FlexGen.
```
Comment (concretevitamin) [inline flexgen/apps/task.yaml]: Add?
```
# References:
#   https://skypilot.readthedocs.io/en/latest/getting-started/quickstart.html
#   https://skypilot.readthedocs.io/en/latest/reference/yaml-spec.html

File: flexgen/apps/task.yaml [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+# Specify the resources required for this job.
+resources:
+  accelerators: T4:1
```
Comment (concretevitamin) [inline flexgen/apps/task.yaml]:
```suggestion
  accelerators: T4:1 # Can replace with other GPU type and count, see `sky show-gpus`
```

File: flexgen/apps/task.yaml [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  instance_type: n1-highmem-32 # On GCP with 1 T4 GPU and more than 200GB of RAM.
```
Comment (concretevitamin) [inline flexgen/apps/task.yaml]: Q: if a user doesn't use GCP (likely), then this YAML may not work? Is it possible to leave out the instance_type (blocked by the memory filter)?

File: flexgen/apps/README.md [hunk 8]
```diff
@@ -12,3 +12,9 @@ Run Massive Multitask Language Understanding (MMLU) scenario.
 ```
 python3 helm_run.py --description mmlu:model=text,subject=abstract_algebra,data_augmentation=canonical --pad-to-seq-len 512 --model facebook/opt-30b --percent 20 80 0 100 0 100 --gpu-batch-size 48 --num-gpu-batches 3 --max-eval-instance 100
 ```
+
+### Run on any cloud with SkyPilot
+Run FlexGen benchmark on any cloud with [SkyPilot](http://skypilot.co).
```
Comment (concretevitamin) [inline flexgen/apps/README.md]: Link to GitHub link?

Add? (link may need to be checked): See detailed instructions in the root [README](../README.md#run-flexgen-on-any-cloud-with-skypilot).

File: README.md [hunk 9]
```diff
@@ -66,9 +66,19 @@ python3 -m flexgen.apps.helm_run --description mmlu:model=text,subject=abstract_
 ```
 Note that only a subset of HELM scenarios is tested. See more tested scenarios [here](flexgen/apps/helm_passed_30b.sh).
 
+### Run FlexGen on any cloud with SkyPilot
+FlexGen benchmark can be launched with [SkyPilot](http://skypilot.co), a tool for launching ML jobs on any cloud.
+You can use a single command below to automatically launch the benchmark on any cloud with SkyPilot, after you setup your cloud account locally (check how to setup SkyPilot [here](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html)).
```
Comment (concretevitamin) [inline README.md]: How about:

First, install SkyPilot and check you have some cloud credentials ([docs](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html#cloud-account-setup)):
```bash
pip install "skypilot[aws,gcp,azure,lambda]"  # pick your clouds
sky check
```
You can now use a single command to automatically launch the benchmark on any cloud:

File: README.md [hunk 10]
```diff
@@ -66,9 +66,19 @@ python3 -m flexgen.apps.helm_run --description mmlu:model=text,subject=abstract_
...
+```
```
Comment (concretevitamin) [inline README.md:75]:
```suggestion
```bash
```

File: flexgen/apps/README.md [hunk 11]
```diff
@@ -12,3 +12,9 @@ Run Massive Multitask Language Understanding (MMLU) scenario.
...
+```
```
Comment (concretevitamin) [inline flexgen/apps/README.md:35]:
```suggestion
```bash
```

File: README.md [hunk 12]
```diff
@@ -66,9 +66,19 @@ python3 -m flexgen.apps.helm_run --description mmlu:model=text,subject=abstract_
...
+sky launch -c flexgen --detach-setup flexgen/apps/task.yaml
```
Comment (concretevitamin) [inline README.md]:
```suggestion
sky launch -c flexgen --detach-setup --down flexgen/apps/task.yaml
```

File: README.md (hunk 3 above)
Comment (concretevitamin) [reply README.md]: I added `--down` and a sentence to explain it. Wdyt? We can also keep the original version of manually running `sky down`. Pros are that the job seems to run pretty long, so people may want to ctrl-c in the middle and manually terminate it. Using autodown showcases a good feature, however.

File: flexgen/apps/task.yaml (hunk 4 above)
Comment (concretevitamin) [inline flexgen/apps/task.yaml]: Consider renaming this file to `skypilot.yaml`? To avoid task vs. job ambiguity.

**Review by concretevitamin — COMMENTED** [review]
Looks great! I can launch this easily.


One UX comment, the progress bar & cursor around the following section seems a bit weird. It doesn't show one line; and the cursor is not at the bottom line. Is it expected?
```
...
(task, pid=19775) Max sequence length: 456, Pad to sequences length: 512
(task, pid=19775) Init weights begin.
(task, pid=19775) Load the pre-trained pytorch weights of opt-30b from huggingface. The downloading and cpu loading can take dozens of minutes. If it seems to get stuck, you can monitor the progress by checking the memory usage of this process.
Downloading (…)l-00007-of-00007.bin: 100%|██████████| 822M/822M [00:11<00:00, 71.0MB/s]
(task, pid=19775) 0002-of-00007.bin:  10%|█         | 1.03G/9.87G [00:11<01:08, 128MB/s]1<01:11, 124MB/s]
(task, pid=19775) 0002-of-00007.bin:  74%|███████▍  | 7.30G/9.87G [00:57<00:17, 150MB/s]57<00:19, 134MB/s]]
(task, pid=19775) 0006-of-00007.bin:  80%|████████  | 7.92G/9.87G [00:57<00:12, 162MB/s]56<00:12, 173MB/s]
(task, pid=19775) Downloading (…)l-00002-of-00007.bin:  74%|███████▍  | 7.34G/9.87G [00:57<00:14, 170MB/s]]
(task, pid=19775) 0002-of-00007.bin:  75%|███████▍  | 7.37G/9.87G [00:57<00:13, 182MB/s]57<00:10, 177MB/s]
Downloading (…)l-00002-of-00007.bin:  76%|███████▌  | 7.49G/9.87G [01:02<02:14, 17.7MB/s]9<00:56, 42.6MB/s]
(task, pid=19775) 0006-of-00007.bin:  82%|████████▏ | 8.10G/9.87G [01:01<01:20, 22.0MB/s]8<00:11, 162MB/s]]
(task, pid=19775) 0005-of-00007.bin:  77%|███████▋  | 7.63G/9.87G [01:01<01:47, 20.7MB/s]8<00:17, 131MB/s]]
Downloading (…)l-00003-of-00007.bin:  83%|████████▎ | 8.16G/9.87G [01:02<01:29, 19.0MB/s]8<00:11, 156MB/s]
(task, pid=19775) Downloading (…)l-00004-of-00007.bin:  50%|████▉     | 4.91G/9.87G [00:57<01:14, 67.0MB/s]
(task, pid=19775) 0004-of-00007.bin:  51%|█████     | 5.02G/9.87G [01:01<03:51, 20.9MB/s]9<01:31, 53.0MB/s]
Downloading (…)l-00001-of-00007.bin:  33%|███▎      | 3.23G/9.79G [01:01<06:27, 16.9MB/s]8<02:25, 45.6MB/s]
```
Maybe it's due to parallel download. Not necessary to fix for this I think.

File: README.md (hunk 1 above)
Comment (Michaelvll) [reply README.md]: Good catch! Done. Thanks!

File: flexgen/apps/task.yaml (hunk 4 above)
Comment (Michaelvll) [reply flexgen/apps/task.yaml]: Done. Thanks!

File: flexgen/apps/task.yaml (hunk 5 above)
Comment (Michaelvll) [reply flexgen/apps/task.yaml]: Done. Thanks!

File: flexgen/apps/task.yaml (hunk 7 above)
Comment (Michaelvll) [reply flexgen/apps/task.yaml]: Yes, leaving out the `instance_type` is blocked by the memory filtering. We cannot just specify `cpus: 32+` because the T4 GPU on AWS with 200+ memory only hold for `g4dn.16xlarge` (64 vCPUs).

File: flexgen/apps/task.yaml (hunk 4 above)
Comment (Michaelvll) [reply flexgen/apps/task.yaml]: Good point! Renamed to `skypilot.yaml`

File: README.md (hunk 10 above)
Comment (Michaelvll) [reply README.md:75]: Done. Thanks!

File: flexgen/apps/README.md (hunk 11 above)
Comment (Michaelvll) [reply flexgen/apps/README.md:35]: Done. Thanks!

File: flexgen/apps/README.md (hunk 8 above)
Comment (Michaelvll) [reply flexgen/apps/README.md]: Good point! Added. Thanks!

File: README.md (hunk 9 above)
Comment (Michaelvll) [reply README.md]: Good point! Replaced. Thanks!

File: README.md (hunk 3 above)
Comment (Michaelvll) [reply README.md]: How about we keep `--down` in the text instead of in the command as following:
```
You can then log into the cluster running the job with `ssh flexgen` for monitoring. Once the job has finished, you can terminate the cluster with `sky down flexgen` or pass in `--down` flag to the command above to have the cluster terminate itself automatically.
```

Reason:
1. With `--down` and if the user detaches from the log, they will never be able to find the log after the cluster is automatically terminated.
2. Adding `--down` makes the launching command longer, which may not look good.

Open to discussions : )

**Comment by Michaelvll** [comment]
> One UX comment, the progress bar & cursor around the following section seems a bit weird. It doesn't show one line; and the cursor is not at the bottom line. Is it expected?

Yea.. that is a pretty annoying problem. It is indeed due to the parallel download, and I don't have a solution right now. Maybe we can leave it for the future.

File: README.md [hunk 13]
```diff
@@ -66,13 +66,20 @@ python3 -m flexgen.apps.helm_run --description mmlu:model=text,subject=abstract_
...
-You can use a single command below to automatically launch the benchmark on any cloud with SkyPilot, after you setup your cloud account locally (check how to setup SkyPilot [here](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html)).
+### Run FlexGen on Any Cloud with SkyPilot
+FlexGen benchmark can be launched with [SkyPilot](https://github.com/skypilot-org/skypilot), a tool for launching ML jobs on any cloud.
+First, install SkyPilot and check you have some cloud credentials ([docs](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html)):
+```bash
+pip install "skypilot[aws,gcp,azure,lambda]"  # pick your clouds
+sky check
 ```
-sky launch -c flexgen --detach-setup flexgen/apps/task.yaml
+You can now use a single command to automatically launch the benchmark on any cloud:
```
Comment (concretevitamin) [inline README.md]:
```suggestion
You can now use a single command to launch the benchmark on any cloud, which automatically finds a region (in the cheapest-price order) with availability for the requested GPUs:
```

File: flexgen/apps/README.md [hunk 14]
```diff
@@ -27,7 +27,16 @@ python3 helm_run.py --description mmlu:model=text,subject=abstract_algebra,data_
...
 ### Run on any cloud with SkyPilot
-Run FlexGen benchmark on any cloud with [SkyPilot](http://skypilot.co).
+FlexGen benchmark can be launched with [SkyPilot](https://github.com/skypilot-org/skypilot), a tool for launching ML jobs on any cloud.
+First, install SkyPilot and check you have some cloud credentials ([docs](https://skypilot.readthedocs.io/en/latest/getting-started/installation.html)):
+```bash
+pip install "skypilot[aws,gcp,azure,lambda]"  # pick your clouds
+sky check
 ```
-sky launch -c flexgen --detach-setup task.yaml
+You can now use a single command to automatically launch the benchmark on any cloud:
```
Comment (concretevitamin) [inline flexgen/apps/README.md]: ditto

File: README.md (hunk 3 above)
Comment (concretevitamin) [reply README.md]: Makes sense. I think either the current version or the original `sky down` version is fine, up to you.

File: flexgen/apps/task.yaml (hunk 7 above)
Comment (concretevitamin) [reply flexgen/apps/task.yaml]: Maybe ok to ship first and see what reports we get. We should expect a non-GCP user to fail at the `sky launch ...` command, however.

**Review by concretevitamin — APPROVED** [review]
